### PR TITLE
Linter: fix structcheck, ineffassign, staticcheck, and errcheck issues

### PR DESCRIPTION
**Pull request by kalensk**
**_Note_**: This PR needs to go in after PRs https://github.com/greenplum-db/gpupgrade/pull/298 and https://github.com/greenplum-db/gpupgrade/pull/300. Be sure to rebase and ensure the lint job on the test pipeline is green before merging.

Continuing from https://github.com/greenplum-db/gpupgrade/pull/298 and https://github.com/greenplum-db/gpupgrade/pull/300 this pulls in @berlin-ab linter work to address additional golangci-lint issues. Take a look a commit at at a time where the last commit adds a lint job to the CI.

To reiterate, we are breaking the linter work into digestible chunks ideally grouping similar issues together into a single commit. This will keep the PRs short and specific to make it easy to review, and to get this merged before the code gets stale and drifts. Further PR's to follow.

[Test Pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:fix_linter_issues_part3) (Note the lint job will fail until PRs https://github.com/greenplum-db/gpupgrade/pull/298 and https://github.com/greenplum-db/gpupgrade/pull/300 are merged, and this PR rebased with master.)

### CODE REVIEW

**Comment by kalensk** [comment]
For the CI _only_, if we want to tie the build process to lint'ing we could add a lint task in the build job, which would make the build job fail for lint errors.

**Comment by jchampio** [comment]
> if we want to tie the build process to lint'ing we could add a lint task in the build job, which would make the build job fail for lint errors.

Right, see the original spike branch for that approach.

File: cli/bash/generate.go [hunk 1]
```diff
@@ -13,5 +14,8 @@ import (
 
 func main() {
 	root := commands.BuildRootCommand()
-	root.GenBashCompletionFile(os.Args[1])
+	err := root.GenBashCompletionFile(os.Args[1])
+	if err != nil {
+		panic(fmt.Sprintf("generating bash-completion.sh: %+v", err))
```
Comment (jchampio) [inline cli/bash/generate.go:19]: `log.Fatalf` is a good combo for simple utilities, to simplify the err-to-string-to-panic.

File: cli/commands/commands.go [hunk 2]
```diff
@@ -390,13 +390,13 @@ After executing, you will need to finalize.`)
 	}
 
 	subInit.Flags().StringVar(&sourceBinDir, "source-bindir", "", "install directory for source gpdb version")
-	subInit.MarkFlagRequired("source-bindir")
+	subInit.MarkFlagRequired("source-bindir") //nolint
```
Comment (jchampio) [inline cli/commands/commands.go:393]: This would be a good reason for a linter config. For now, `nolint` is fine.

**Comment by kalensk** [comment]
> Right, see the original spike branch for that approach.

In https://github.com/greenplum-db/gpupgrade/pull/210 I am not seeing where the possibility of ting the linter to the build job _only_ on the CI is discussed.

File: cli/bash/generate.go (hunk 1 above)
Comment (kalensk) [reply cli/bash/generate.go]: fixed.

**Pull request merged**